### PR TITLE
Continuation of casting implementation

### PR DESCRIPTION
**Pull request by EtienneDesticourt**
TODO still:
Write tests for events
Implement after_cast_end event
Add callback parameters instead of default nil
Add recharge_callback call

Questions:
Is the diagram up to date?
What's the purpose of after_cast_timer?
What's the purpose of the recharge callback ?
Why do we only check if the player is casting when the skill is not instantaneous?

### CODE REVIEW

**Comment by ephe-meral** [comment]
- the diagram should be up to date yeah... why do you ask, is anything suspicious?
- after_cast is the short timespan after casting a skill (usually around 250ms afair) that you cannot do anything else (this happens parallel to recharge)
- recharge callback and the other callbacks might not be needed, depending on what we decide to do upon successful casting... it used to be necessary inform the player via a channel that the casting is done. but now we can either send a message to the channel process, or we can watch for changes in the entity from the channel, we should discuss both options :)
- instantaneous skill you can trigger while you cast something, no? or is that different?

**Comment by ephe-meral** [comment]
also, this looks very good! thanks for the work :)

File: lib/entice/logic/casting.ex [hunk 1]
```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
...
+  do: Entity.remove_behaviour(entity_id, Casting)
 
   @doc "Deals with timing and thus might fail. Should be called by the Skillbar"
   def cast_skill(entity, skill) when is_atom(skill),
-  do: entity |> Entity.call_behaviour({:skill_cast_start, %{target: nil, skill: skill}})
+  do: entity |> Entity.call_behaviour(Casting.Behaviour, {:cast_start, nil, nil, %{target: nil, skill: skill}})
 
   def cast_skill(entity, target, skill) when is_atom(skill),
-  do: entity |> Entity.call_behaviour({:skill_cast_start, %{target: target, skill: skill}})
+  do: Entity.call_behaviour(entity, Casting.Behaviour, {:cast_start, nil, nil, %{target: target, skill: skill}})
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex]: lets either do `entity |> ...` or the parameter version in both methods

File: lib/entice/logic/casting.ex [hunk 2]
```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
...
-              {:ok, :ok, entity}
-            result -> raise "Incorrect skill casting result:\nSkill: #{skill.id}\nResult: #{result}"
-          end
+    def init(entity, %Casting{} = casting),
+    do: {:ok, entity |> put_attribute(casting)}
+
+    def terminate(_reason, entity),
+    do: {:ok, entity |> remove_attribute(Casting)}
+
+    def handle_call({:cast_start, cast_callback, recharge_callback, %{target: _target, skill: skill}}, %Entity{attributes: %{
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex]: lets fail gracefully instead of crashing by implementing a catch all default handle_call below this method - it might well be that the entity doesn't have energy

File: test/entice/logic/casting_test.exs [hunk 3]
```diff
@@ -0,0 +1,70 @@
+defmodule Entice.Logic.CastingTest do
```
Comment (ephe-meral) [inline test/entice/logic/casting_test.exs:1]: niiiiiiiiiice :heart: we definitely need testing for this yeah

File: lib/entice/logic/casting.ex [hunk 4]
```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
...
+      recharge_timers = Map.remove(entity.recharge_timers, skill)
+      {:ok, entity |> update_attribute(Casting, fn s -> %Casting{s | recharge_timers: recharge_timers} end)}
+    end
+
+    def handle_event({:after_cast_end, _skill, _recharge_callback}, entity) do
+      {:ok, entity |> update_attribute(Casting, fn s -> %Casting{s | after_cast_timer: nil} end)}
+    end
+
+    defp reduce_mana(entity, new_mana) do
+      Entity.update_attribute(entity, Energy, fn e -> %Energy{e | mana: new_mana} end)
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex]: please don't call the Entity.[something] api from within a behaviour since it's an additional indirection when we can simply update the attributes from within this

File: lib/entice/logic/casting.ex [hunk 5]
```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
...
+
+    def handle_call({:cast_start, cast_callback, recharge_callback, %{target: _target, skill: skill}}, %Entity{attributes: %{
+      Casting => %Casting{recharge_timers: _recharge_timers},
+      Energy => %Energy{mana: mana}}} = entity) do
+
+      response = can_cast?(skill, entity)
+      entity = case response do
+        {:ok, _} ->
+          entity
+          |> reduce_mana(mana - skill.energy_cost)
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex:49]: if possible, this should be the last statement in this case branch and it should return the new entity state, as i mention below in reduce_mana

File: lib/entice/logic/casting.ex [hunk 6]
```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
...
-    casting_timer: nil,
+    casting_timer: nil,  #Why not call this cast_timer if skill has a cast_time ? or vice versa
     after_cast_timer: nil,
     recharge_timers: %{})
 
-
   def register(entity_id),
-  do: Entity.put_behaviour(entity_id, Casting.Behaviour, [])
-
+  do: Entity.put_behaviour(entity_id, Casting.Behaviour, %Casting{})
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex:20]: thats such a great idea to initialize the behaviour with the attribute we want to add :D
i like that... maybe we should do that with other stuff as well some time in the future
also enables us to initialize the entities in a more simple fashion, e.g. for testing

File: lib/entice/logic/casting.ex [hunk 7]
```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
...
+    @doc "This event triggers when a skill's recharge period ends, it resets the recharge timer for the skill."
+    def handle_event({:recharge_end, skill, _recharge_callback}, entity) do
+      recharge_timers = Map.remove(entity.recharge_timers, skill)
+      {:ok, entity |> update_attribute(Casting, fn s -> %Casting{s | recharge_timers: recharge_timers} end)}
+    end
+
+    def handle_event({:after_cast_end, _skill, _recharge_callback}, entity) do
+      {:ok, entity |> update_attribute(Casting, fn s -> %Casting{s | after_cast_timer: nil} end)}
+    end
+
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex:80]: before the private stuff comes we still need to make sure the attribute is remove when the behaviour is terminated, check another behaviour for how to do that :)

File: lib/entice/logic/casting.ex [hunk 8]
```diff
@@ -8,51 +8,131 @@ defmodule Entice.Logic.Casting do
   alias Entice.Logic.Casting
   alias Entice.Logic.Vitals.Energy
   alias Entice.Entity
-  alias Entice.Entity.Coordination
+  use Pipe
 
 
   defstruct(
-    casting_timer: nil,
+    casting_timer: nil,  #Why not call this cast_timer if skill has a cast_time ? or vice versa
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex]: true that - what do you think works better?

File: lib/entice/logic/casting.ex [hunk 9]
```diff
@@ -8,51 +8,131 @@ defmodule Entice.Logic.Casting do
...
+    def handle_event({:recharge_end, skill, recharge_callback}, entity) do
+      recharge_callback.(skill)
+      {:ok, entity |> update_attribute(Casting, fn c -> %Casting{c | recharge_timers: c.recharge_timers |> Map.delete(skill)} end)}
+    end
+
+    def handle_event({:after_cast_end}, entity) do
+      {:ok, entity |> update_attribute(Casting, fn c -> %Casting{c | after_cast_timer: nil} end)}
+    end
+
+    def handle_event({:test_message_1}, entity) do
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex]: not needed ;)

File: lib/entice/logic/casting.ex [hunk 10]
```diff
@@ -8,51 +8,131 @@ defmodule Entice.Logic.Casting do
...
+    defp enough_energy?({:ok, skill}, dmana) when dmana > 0, do: {:ok, skill}
+    defp enough_energy?({:ok, _skill}, _dmana), do: {:error, :not_enough_energy}
+
+    defp not_recharging?({:ok, skill}, nil = _recharge_timer), do: {:ok, skill}
+    defp not_recharging?(_skill, _recharge_timer), do: {:error, :still_recharging}
+
+    defp not_casting?({:ok, skill}, nil = _casting_timer, nil = _after_cast_timer), do: {:ok, skill}
+    defp not_casting?(_skill, _casting_timer, _after_cast_timer), do: {:error, :still_casting}
+
+    defp can_cast?(skill,  %Entity{attributes: %{
```
Comment (ephe-meral) [inline lib/entice/logic/casting.ex:106]: this is just concerning code structure, but personally i think its more clear if we put functions in and order like:
- 1) some function
- some sub-function that is used by 1
- another sub function used by 1
- 2) some other function
- a function used by 2)

etc.

tl;dr: can you move the helpers from above used by this below it?

File: test/entice/logic/casting_test.exs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+  setup do
+    {:ok, entity_id, _pid} = Entity.start
+    Casting.register(entity_id)
+    Spy.register(entity_id, self)
+    Entity.put_attribute(entity_id, %Energy{mana: 50})
+    {:ok, [entity_id: entity_id]}
+  end
+
+  @tag id: 0, casting: true
```
Comment (ephe-meral) [inline test/entice/logic/casting_test.exs:18]: do we use the tags for something?

File: lib/entice/logic/casting.ex (hunk 8 above)
Comment (EtienneDesticourt) [inline lib/entice/logic/casting.ex]: Ive changed it to cast_timer

File: lib/entice/logic/casting.ex (hunk 6 above)
Comment (EtienneDesticourt) [inline lib/entice/logic/casting.ex:20]: I just copied this from skillbar ;)

File: test/entice/logic/casting_test.exs (hunk 11 above)
Comment (EtienneDesticourt) [inline test/entice/logic/casting_test.exs:18]: id we should use for setup but Im too lazy to do it. 
casting I use to do mix test --only casting it's faster for repeat tests. We could do this via file path but it's cleaner that way.

Comment (ephe-meral) [inline test/entice/logic/casting_test.exs:18]: alright cool, good to know ;)

File: lib/entice/logic/casting.ex (hunk 6 above)
Comment (ephe-meral) [inline lib/entice/logic/casting.ex:20]: then i forgot about this :D

**Comment by ephe-meral** [comment]
:heart:
